### PR TITLE
feat(glm52): DSA indexer kernel ops — cache, topk, slots, hadamard, DeepGEMM MQA logits

### PR DESCRIPTION
**Pull request by xiaguan**
## Summary

PR2 of the GLM5.2 DSA (DeepSeek Sparse Attention) indexer chain ([roadmap](docs/models/glm52/dp1-ep8-decode-plan.md)). Implements all 6 kernel ops needed to replace PR1's full top-k with real sparse top-k=2048 selection.

## Kernel ops

| op | backend | hand-written? |
|---|---|---|
| `glm52_indexer_k_quant_and_cache` | fp8 per-128-group quant + scatter into paged cache | yes (cherry-pick) |
| `glm52_indexer_k_gather_quant_cache` | gather fp8 KV blocks from paged cache | yes (cherry-pick) |
| `glm52_indexer_local_topk_to_slots` | int32 block-table remap (local topk → global slots) | yes (ported from TokenSpeed Triton) |
| `glm52_indexer_hadamard_bf16` | naive radix Hadamard for head_dim=128 | yes |
| `glm52_flashinfer_topk_2048` | FlashInfer `TopKDispatch` K=2048, deterministic | no (vendored) |
| `glm52_deepgemm_paged_mqa_logits` | DeepGEMM `sm90_fp8_paged_mqa_logits` JIT | no (vendored, torch-free wrapper) |

## DeepGEMM torch-free approach

The DeepGEMM paged MQA logits wrapper is the first real DeepGEMM JIT kernel call in the codebase. Instead of linking libtorch, 7 vendored DeepGEMM headers are patched with `#ifdef DG_NO_TORCH` conditional compilation (~150 lines of guards):

- `compatibility.hpp` / `math.hpp` — guard `<torch/version.h>` / `<torch/python.h>`
- `device_runtime.hpp` — guard cuBLASLt workspace + `ATen/CUDAContext.h`
- `compiler.hpp` — guard vestigial `ATen` include; add `<optional>`
- `kernel_runtime.hpp` — accept `cudaStream_t` param instead of `at::cuda::getCurrentCUDAStream()`
- `runtime_utils.hpp` — add `make_tma_2d/3d_desc_raw` (raw `void*` + `DgDtype` enum) + guard `cute::UMMA::Major` / `GemmType` / `at::ScalarType` functions
- `sm90_fp8_mqa_logits.hpp` — `at::ScalarType` → `DgDtype`; guard torch free functions

The C ABI wrapper (`csrc/glm52/glm52_deepgemm_mqa.cu`) calls `SM90FP8PagedMQALogitsRuntime` + `SM90PagedMQALogitsMetadataRuntime` directly with raw pointers and manually-constructed TMA descriptors.

Investigated FlashInfer group GEMM as alternative — rejected (no paged KV cache support, no per-128-group fp8 scales on SM90).

## Test results

All 6 smoke tests pass on H200 (jz38):

```
test test_indexer_cache_round_trip ... ok
test test_local_topk_to_slots_basic ... ok
test test_local_topk_to_slots_invalid ... ok
test test_hadamard_correctness ... ok
test test_flashinfer_topk_basic ... ok
test test_deepgemm_paged_mqa_launch ... ok
test result: ok. 6 passed; 0 failed
```

DeepGEMM test requires `OPENINFER_DEEPGEMM_ROOT` + `CUDA_HOME` env vars (JIT compile at runtime, cubin cached at `~/.deep_gemm/`).

## Not in this PR

- Model crate `indexer.rs` forward wiring (PR3+)
- Oracle gate (deferred — same fixture-pipeline blocker as PR1)
- Dao-AILab fast-hadamard-transform port (follow-up if naive version is ncu bottleneck)

## Dev doc

See `docs/models/glm52/dsa-indexer.md` for full design notes, execution log, and DeepGEMM patch details.

### CODE REVIEW

**Comment by xiaguan** [comment]
@codex review

**Comment by xiaguan** [comment]
@codex review

**Comment by xiaguan** [comment]
@codex review

**Comment by xiaguan** [comment]
@codex review
